File: CSharp/UniRecog.cs
using System;
using System.IO;
using System.Threading;

class UniRecog
{
    // UniMRCP client root directory
    static string ROOT_DIR = "../../../../trunk";
    // UniMRCP profile to use for communication with server
    const string MRCP_PROFILE = "uni1";
    // File to write Recogesized text to (ROOT_DIR/data/PCM_OUT_FILE)
    const string PCM_OUT_FILE = "UniRecog.pcm";

    // Types of streaming
    private enum StreamType
    {
        FRAME_BY_FRAME,
        BUFFERED,
        MEMORY,
        FILE
    }

    // Choose type of streaming to demonstrate
    private const StreamType streamType = StreamType.FILE;

    static int err = 0;

    // User-defined logger
    class UniRecogLogger : UniMRCPLogger
    {
        // Log messages go here
        public override bool Log(string file, uint line, UniMRCPLogPriority priority, string message)
        {
            Console.WriteLine("  {0}", message);
            return true;
        }
    }


    class UniRecogSession : UniMRCPClientSession
    {
        public UniRecogSession(UniMRCPClient client, string profile) :
            base(client, profile)
        { }

        public override bool OnTerminate(UniMRCPSigStatusCode status)
        {
            Console.WriteLine("Session terminated with code {0}", status);
            return true;
        }

        public override bool OnTerminateEvent()
        {
            Console.WriteLine("Session terminated unexpectedly");
            return true;
        }
    }


    class UniRecogStreamFrames : UniMRCPStreamRx
    {
        // Set to true upon RECOGNIZE's IN-PROGRESS
        public bool started = false;
        // Output file descriptor
        private FileStream f = null;
        // Buffer to copy audio data to
        private byte[] buf = null;

        public UniRecogStreamFrames(string pcmFile)
        {
            f = new FileStream(pcmFile, FileMode.Open, FileAccess.Read);
        }

        // Called when an audio frame is needed
        public override bool ReadFrame()
        {
            if (!started)
                return false;
            uint frameSize = GetDataSize();
            if (frameSize <= 0)
                return true;
            if (buf == null)
                buf = new byte[frameSize];
            int bytesRead = f.Read(buf, 0, (int)frameSize);
            if (bytesRead > 0)
            {
                for (int i = bytesRead; i < buf.Length; ++i)
                    buf[i] = 0;
                SetData(buf);
                return true;
            }
            return false;
        }
    }


    class UniRecogTermination : UniMRCPAudioTermination
    {
        public UniMRCPStreamRx stream;

        public UniRecogTermination(UniMRCPClientSession sess, string pcmFile) :
            base(sess)
        {
            switch (streamType)
            {
                case StreamType.FRAME_BY_FRAME:
                    stream = new UniRecogStreamFrames(pcmFile);
                    break;
                case StreamType.BUFFERED:
                    stream = new UniMRCPStreamRxBuffered();
                    break;
                case StreamType.MEMORY:
                    byte[] data = File.ReadAllBytes(pcmFile);
                    stream = new UniMRCPStreamRxMemory(data, true, UniMRCPStreamRxMemory.StreamRxMemoryEnd.SRM_NOTHING, true);
                    break;
                case StreamType.FILE:
                    stream = new UniMRCPStreamRxFile(pcmFile, 0, UniMRCPStreamRxMemory.StreamRxMemoryEnd.SRM_NOTHING, true);
                    break;
            }
        }

        public override UniMRCPStreamRx OnStreamOpenRx(bool enabled, byte payload_type, string name, string format, byte channels, uint freq)
        {
            // Configure incoing stream here
            return stream;
        }
    }


    class UniRecogChannel : UniMRCPRecognizerChannel
    {
        private UniRecogTermination term;
        private Semaphore sem;
        private string grammarFile;
        private string inputFile;

        public UniRecogChannel(UniRecogSession sess, UniRecogTermination term, Semaphore sem, string grammarFile, string inputFile) :
            base(sess, term)
        {
            this.term = term;
            this.sem = sem;
            this.grammarFile = grammarFile;
            this.inputFile = inputFile;
        }

        // Shorthand for graceful fail: Write message, release semaphore and return false
        private bool Fail(string msg)
        {
            Console.WriteLine(msg);
            err = 1;
            sem.Release();
            return false;
        }

        // MRCP connection established, start communication
        public override bool OnAdd(UniMRCPSigStatusCode status)
        {
            if (status != UniMRCPSigStatusCode.SUCCESS)
                return Fail(String.Format("Failed to add channel: {0}", status));
            // Load grammar content from file
            byte[] content = File.ReadAllBytes(grammarFile);
            // Start processing here
            UniMRCPRecognizerMessage msg = CreateMessage(UniMRCPRecognizerMethod.RECOGNIZER_RECOGNIZE);
            msg.content_type = "application/grammar+xml";
            msg.SetBody(content);
            return msg.Send();
        }

        // Response or event from the server arrived
        public override bool OnMessageReceive(UniMRCPRecognizerMessage message)
        {
            // Analyze message, update your application state and reply messages here
            if (message.GetMsgType() == UniMRCPMessageType.RESPONSE)
            {
                if (message.GetStatusCode() != UniMRCPStatusCode.CODE_SUCCESS)
                    return Fail(String.Format("RECOGNIZE request failed: {0}", message.GetStatusCode()));
                if (message.GetRequestState() != UniMRCPRequestState.STATE_INPROGRESS)
                    return Fail(String.Format("Failed to start RECOGNIZE processing"));
                // Start streaming audio from the file
                StartStreaming();
                return true;  // Does not actually matter
            }
            if (message.GetMsgType() != UniMRCPMessageType.EVENT)
                return Fail("Unexpected message from the server");
            if (message.GetEventID() == UniMRCPRecognizerEvent.RECOGNIZER_START_OF_INPUT_EVENT)
            {
                Console.WriteLine("Start of input detected");
                return true;  // Does not actually matter
            }
            if (message.GetEventID() == UniMRCPRecognizerEvent.RECOGNIZER_RECOGNITION_COMPLETE)
            {
                Console.WriteLine(String.Format("Recognition complete: {0} {1}", message.completion_cause, message.completion_reason));
                sem.Release();
                return true;  // Does not actually matter
            }
            return Fail("Unknown message received");
        }

        private void StartStreaming()
        {
            switch (streamType)
            {
                case StreamType.FRAME_BY_FRAME:
                    ((UniRecogStreamFrames)term.stream).started = true;
                    break;
                case StreamType.BUFFERED:
                    try
                    {
                        byte[] data = File.ReadAllBytes(inputFile);
                        ((UniMRCPStreamRxBuffered)term.stream).AddData(data);
                    }
                    catch (IOException)
                    {
                    }
                    break;
                case StreamType.MEMORY:
                case StreamType.FILE:
                    ((UniMRCPStreamRxMemory)term.stream).SetPaused(false);
                    break;
            }
        }
    }


    static void Main(string[] argv)
    {
        if (argv.Length < 2)
        {
            Console.WriteLine("Usage:");
            Console.WriteLine("\tUniRecog \"/path/to/grammar/file\" \"/path/to/input/file\"");
            Environment.Exit(1);
        }
        // Just detect various directory layout constellations
        if (!Directory.Exists(ROOT_DIR))
            ROOT_DIR = "../../../../UniMRCP";
        if (!Directory.Exists(ROOT_DIR))
            ROOT_DIR = "../../../../unimrcp";
        string grammarFile = argv[0].Trim();
        string inputFile = argv[1].Trim();

        ushort major, minor, patch;
        UniMRCPClient.WrapperVersion(out major, out minor, out patch);
        Console.WriteLine("This is a sample C# UniMRCP client recognizer scenario.");
        Console.WriteLine("Wrapper version: {0}.{1}.{2}", major, minor, patch);
        Console.WriteLine("Use client configuration from {0}/conf/unimrcpclient.xml", ROOT_DIR);
        Console.WriteLine("Use profile {0}", MRCP_PROFILE);
        Console.WriteLine("Use recognition grammar: `{0}'", grammarFile);
        Console.WriteLine("Use recognition input file: `{0}'", inputFile);
        Console.WriteLine();
        Console.WriteLine("Press enter to start the session...");
        Console.ReadKey();

        UniRecogLogger logger = new UniRecogLogger();
        try
        {
            // Initialize platform first
            UniMRCPClient.StaticInitialize(logger, UniMRCPLogPriority.INFO);
        }
        catch (ApplicationException ex)
        {
            Console.WriteLine("Unable to initialize platform: " + ex.Message);
            Environment.Exit(1);
        }

        UniMRCPClient client = null;
        UniRecogSession sess = null;
        UniRecogChannel chan = null;
        try
        {
            // Create and start client in a root dir
            client = new UniMRCPClient(ROOT_DIR, true);
            // Create a session using MRCP profile MRCP_PROFILE
            sess = new UniRecogSession(client, MRCP_PROFILE);
            // Create audio termination with capabilities
            UniRecogTermination term = new UniRecogTermination(sess, inputFile);
            term.AddCapability("LPCM", UniMRCPSampleRate.SAMPLE_RATE_8000);
            // Semaphore to wait for completion
            Semaphore sem = new Semaphore(0, 1);
            // Add signaling channel (and start processing in OnAdd method
            chan = new UniRecogChannel(sess, term, sem, grammarFile, inputFile);

            // Now wait until the processing finishes
            sem.WaitOne();
        }
        catch (ApplicationException ex)
        {
            err = 1;
            Console.WriteLine("An error occured: " + ex.Message);
        }

        // Ensure correct destruction order (do not let GC decide)
        if (chan != null) chan.Dispose();
        if (sess != null) sess.Dispose();
        if (client != null) client.Dispose();

        UniMRCPClient.StaticDeinitialize();
        Console.WriteLine("Program finished, memory released. Press enter to exit.");
        Console.ReadKey();
        Environment.Exit(err);
    }
}

File: Python/UniRecog.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
import os
import sys
import threading

# UniMRCP client root directory
ROOT_DIR = "../../../../trunk"
# Just detect various directory layout constellations
if not os.path.isdir(ROOT_DIR):
    ROOT_DIR = "../../../../UniMRCP"
if not os.path.isdir(ROOT_DIR):
    ROOT_DIR = "../../../../unimrcp"
# UniMRCP profile to use for communication with server
MRCP_PROFILE = "uni1"

# Types of streaming
ST_FRAME_BY_FRAME = 0
ST_BUFFERED = 1
ST_MEMORY = 2
ST_FILE = 3

# Choose type of streaming to demonstrate
streamType = ST_FILE


# Import UniMRCP symbols
from UniMRCP import *


# User-defined logger
class UniRecogLogger(UniMRCPLogger):
    def __init__(self):
        super(UniRecogLogger, self).__init__()

    # Log messages go here
    def Log(self, file, line, prio, message):
        print "  %s\n" % message,  # Ensure atomic logging, do not intermix
        return True


class UniRecogSession(UniMRCPClientSession):
    def __init__(self, client, profile):
        super(UniRecogSession, self).__init__(client, profile)

    def OnTerminate(self, status):
        print "Session terminated with code", status
        return True

    def OnTerminateEvent(self):
        print "Session terminated unexpectedly"
        return True


class UniRecogStreamFrames(UniMRCPStreamRx):
    def __init__(self, pcmfile):
        super(UniRecogStreamFrames, self).__init__()
        # Set to True upon RECOGNIZE's IN-PROGRESS
        self.started = False
        # Input file descriptor
        self.f = open(pcmfile, 'rb')

    # Called when an audio frame is needed
    def ReadFrame(self):
        if not self.started:
            return False
        frameSize = self.GetDataSize()
        if frameSize:
            buf = self.f.read(frameSize)
            if len(buf):
                self.SetData(buf)
                return True
        return False


class UniRecogTermination(UniMRCPAudioTermination):
    def __init__(self, sess, pcmfile):
        super(UniRecogTermination, self).__init__(sess)
        if streamType == ST_FRAME_BY_FRAME:
            self.stream = UniRecogStreamFrames(pcmfile)
        elif streamType == ST_BUFFERED:
            self.stream = UniMRCPStreamRxBuffered()
        elif streamType == ST_MEMORY:
            data = open(pcmfile, "rb").read()
            self.stream = UniMRCPStreamRxMemory(data, True, UniMRCPStreamRxMemory.SRM_NOTHING, True)
        elif streamType == ST_FILE:
            self.stream = UniMRCPStreamRxFile(pcmfile, 0, SRM_NOTHING, True)

    def OnStreamOpenRx(self, enabled, payload_type, name, format, channels, freq):
        # Configure outgoing stream here
        return self.stream


class UniRecogChannel(UniMRCPRecognizerChannel):
    def __init__(self, sess, term, sem, grammarfile, inputfile):
        super(UniRecogChannel, self).__init__(sess, term)
        self.sess = sess
        self.term = term
        self.sem = sem
        self.grammarfile = grammarfile
        self.inputfile = inputfile

    # Shorthand for graceful fail: Write message, release semaphore and return False
    def Fail(self, msg):
        global err
        print msg
        err = 1
        self.sem.release()
        return False

    # MRCP connection established, start communication
    def OnAdd(self, status):
        if status != MRCP_SIG_STATUS_CODE_SUCCESS:
            return self.Fail("Failed to add channel: %d" % status)
        # Load grammar content from file
        content = open(grammarfile, "rb").read()
        # Start processing here
        msg = self.CreateMessage(RECOGNIZER_RECOGNIZE)
        msg.content_type = "application/grammar+xml"
        msg.SetBody(content)
        return msg.Send()

    # Response or event from the server arrived
    def OnMessageReceive(self, message):
        # Analyze message, update your application state and reply messages here
        if message.GetMsgType() == MRCP_MESSAGE_TYPE_RESPONSE:
            if message.GetStatusCode() != MRCP_STATUS_CODE_SUCCESS:
                return self.Fail("RECOGNIZE request failed: %d" % message.GetStatusCode())
            if message.GetRequestState() != MRCP_REQUEST_STATE_INPROGRESS:
                return self.Fail("Failed to start RECOGNIZE processing")
            # Start streaming audio from the file
            self.StartStreaming()
            return True  # Does not actually matter
        if message.GetMsgType() != MRCP_MESSAGE_TYPE_EVENT:
            return self.Fail("Unexpected message from the server")
        if message.GetEventID() == RECOGNIZER_START_OF_INPUT_EVENT:
            print "Start of input detected"
            return True
        if message.GetEventID() == RECOGNIZER_RECOGNITION_COMPLETE:
            print "Recognition complete:", message.completion_cause, message.completion_reason
            self.sem.release()
            return True  # Does not actually matter
        return self.Fail("Unknown message received")

    def StartStreaming(self):
        if streamType == ST_FRAME_BY_FRAME:
            self.term.stream.started = True
        elif streamType == ST_BUFFERED:
            data = open(self.inputfile, "rb").read()
            self.term.stream.AddData(data)
        elif streamType == ST_MEMORY:
            self.term.stream.SetPaused(False)
        elif streamType == ST_FILE:
            self.term.stream.SetPaused(False)


if len(sys.argv) < 3:
    print "Usage:"
    print '\t%s "/path/to/grammar/file" "/path/to/input/file"' % sys.argv[0]
    sys.exit(1)
grammarfile = sys.argv[1]
inputfile = sys.argv[2]

print "This is a sample Python UniMRCP client recognizer scenario."
print "Wrapper version: {0}.{1}.{2}".format(*UniMRCPClient.WrapperVersion())
print "Use client configuration from %s/conf/unimrcpclient.xml" % ROOT_DIR
print "Use profile %s" % MRCP_PROFILE
print "Use recognition grammar: `%s'" % grammarfile
print "Use recognition input file: %s" % inputfile
print
print "Press enter to start the session..."
sys.stdin.read(1)

logger = UniRecogLogger()
try:
    # Initialize platform first
    UniMRCPClient.StaticInitialize(logger, APT_PRIO_DEBUG)
except RuntimeError as ex:
    print "Unable to initialize platform:", ex
    sys.exit(1)

err = 0
client = None
sess = None
term = None
chan = None
try:
    # Create and start client in a root dir
    client = UniMRCPClient(ROOT_DIR, True)
    # Create a session using MRCP profile MRCP_PROFILE
    sess = UniRecogSession(client, MRCP_PROFILE)
    # Create audio termination with capabilities
    term = UniRecogTermination(sess, inputfile)
    term.AddCapability("LPCM", SAMPLE_RATE_8000)
    # Semaphore to wait for completion
    sem = threading.Semaphore(0)
    # Add signaling channel (and start processing in OnAdd method)
    chan = UniRecogChannel(sess, term, sem, grammarfile, inputfile)

    # Now wait until the processing finishes
    sem.acquire()
except Exception as ex:
    err = 1
    print "An error occured:", ex

# Ensure correct destruction order (do not let GC decide)
chan = None
sess = None
client = None

UniMRCPClient.StaticDeinitialize()
print "Program finished, memory released. Press enter to exit."
sys.stdin.read(1)
sys.exit(err)
